cheatsheets/Password_Storage_Cheat_Sheet.md

```diff
@@ -1,567 +1,151 @@
 
-Example `protect()` pseudo-code follows:
+When choosing a work factor, a balance needs to be struck between security and performance. Higher work factors will make the hashes more difficult for an attacker to crack, but will also make the process of verifying a login attempt slower. If the work factor is too high, this may degrade the performance of the application, and could also be used by an attacker to carry out a denial of service attack by making a large number of login attempts to exhaust server's CPU.
 
-```text
-return [salt] + HMAC-SHA-256([key], [salt] + [credential]);
-```
+There is no golden rule for the ideal work factor - it will depend on the performance of the server and the number of users on the application. Determining the optimal work factor will require experimentation on the specific server(s) used by the application. As a general rule, the calculating a hash should take less than one second, although on higher traffic sites it should be significantly less than this.
 
-Upholding security improvement over (solely) salted schemes relies on proper key management.
+### Upgrading the Work Factor
 
-## Design password storage assuming eventual compromise
+One key advantage of having a work factor is that it can be increased over time as hardware becomes more powerful and cheaper. Taking Moore's Law (i.e, that computational power at a given price point doubles every eighteen months) as a rough approximation, this means that the work factor should be increased by 1 every eighteen months.
 
-The frequency and ease with which threats steal protected credentials demands "design for failure". Having detected theft, a credential storage scheme must support continued operation by marking credential data as compromised. It's also critical to engage alternative credential validation workflows as follows:
+The most common approach to upgrading the work factor is to wait until the user next authenticates, and then to re-hash their password with the new work factor. This means that different hashes will have different work factors, and may result in hashes never being upgraded if the user doesn't log back in to the application. In some cases, it may be appropriate to remove the older password hashes and require users to reset their passwords next time they need to login, in order to avoid storing older and less secure hashes.
 
-1. Protect the user's account
-    1. Invalidate authentication 'shortcuts' by disallowing login without 2nd factors, secret questions or some other form of strong authentication.
-    2. Disallow changes to user accounts such as editing secret questions and changing account multi-factor configuration settings.
+In some cases, it may be possible to increase the work factor of the hashes without the original password, although this is not supported by common hashing algorithms such as Bcrypt and PBKDF2.
```

Comment:
You can certainly do this with something like MD5, but as Bcrypt/etc need the salt to calculate the hash, and the salt is stored in the output, I don't think this would work for Bcrypt? Unless you did something odd like splitting out the hash into its own database field and do you own slightly custom code around Bcrypt, but that seems like a bad idea.

Reply:
As far as I'm aware, you can't increase the work factor of existing Bcrypt hashes without the original password?

Reply:
I got your point. I am with keeping it to that simple note that you set. I was mainly answering about upgrading the work factors in a manual way.